File: 1/2020.10.04/169.Majority Element/java/solution.java
class Solution {
    public int majorityElement(int[] nums) {
        // sort
        Arrays.sort(nums);
        return nums[nums.length/2];
    }

    public int majorityElement(int[] nums) {
        // hash table
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
            if (map.get(num) > nums.length/2) {
                return num;
            }
        }
        throw new RuntimeException("no majority element found.");
    }

    public int majorityElement(int[] nums) {
        // boyer-moore 投票算法
        int candidate = 0, count = 0;
        for (int num : nums) {
            if (count == 0) {
                candidate = num;
            }
            count += candidate == num ? 1 : -1;
        }
        return candidate;
    }
}

File: 1/2020.10.04/169.Majority Element/python/solution.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        # boyer-moore 投票算法
        count = 0
        for num in nums:
            if count == 0:
                candidate = num
            count += 1 if num == candidate else -1
        return candidate

    def majorityElement(self, nums: List[int]) -> int:
        # sort
        nums = sorted(nums)
        return nums[len(nums) // 2]

    def majorityElement(self, nums: List[int]) -> int:
        # hash table
        d = {}
        for num in nums:
            d[num] = d.get(num, 0) + 1
            if d[num] > len(nums) // 2:
                return num
